### ISSUE TITLE
synchronize: cannot use rsync's --inplace option, because --delay-updates is hardcoded

### ISSUE THREAD

**Issue by breun** [ansible/ansible-modules-core#727]
It seems the synchronize module calls rsync with the `--delay-updates` option by default. `man rsync` says the following about `--delay-updates`:

> This  option  uses  more memory on the receiving side (one bit per file transferred) and also requires enough free disk space on the receiving side to hold an additional copy of all the updated files.

We want to synchronize large directories containing databases on servers that may not have enough disk space to hold an additional copy of all the updates files. Normally one could use rsync's `--inplace` option in a case like this, but sadly using `rsync_opts=--inplace` with Ansible's synchronize module doesn't work, because it uses `--delay-updates` by default, which is incompatible with `--inplace`.

**Comment by EqO** [comment]
+1 for allowing `--inplace` option

**Comment by 0utsider** [comment]
+1 for --inplace option !

**Comment by nowak-ninja** [comment]
+1

**Comment by evilr00t** [comment]
+1

**Comment by seses** [comment]
+1

**Comment by aosmialowski** [comment]
+1

**Comment by mchybowski-rtbh** [comment]
+1

**Comment by maciejstromich** [comment]
+1

**Comment by gprochnicki** [comment]
+1

**Comment by frogu** [comment]
+1

**Comment by krzysztof-magosa** [comment]
+1

**Comment by kporembinski** [comment]
+1

**Comment by keiosweb** [comment]
+1

**Comment by ad-m** [comment]
+1

**Comment by mploski** [comment]
+1

**Comment by buker** [comment]
+1

**Comment by teklakct** [comment]
+1

**Comment by michalbagrowski** [comment]
+1

**Comment by smereczynski** [comment]
+1

**Comment by pniederlag** [comment]
--delay-updates simply breaks the rsync on large filesets for me, so +1 for giving a chance to disable it

**Comment by krzysztof-magosa** [comment]
+1

**Comment by xiongchiamiov** [comment]
Aside from the memory and disk space requirements, it also just makes the whole process much slower:  on an example directory I have with 2.2 million files (yes, I know), the command as it comes from Ansible takes 5.5 minutes just to build the file list upfront, while removing `--delay-updates` allows rsync to use [its incremental file list](https://rsync.samba.org/FAQ.html#4), which resulted in the _entire_ copy operation taking only 3.5 minutes.

**Comment by xiongchiamiov** [comment]
Also (sorry for the spam), [the copy module](https://docs.ansible.com/ansible/copy_module.html) notes that its

> recursively copy facility does not scale to lots (>hundreds) of files. For alternative, see synchronize module, which is a wrapper around rsync.

which means there isn't really a good option right now for transferring lots of files, other than crafting a `command` invocation (a design decision I'm often grateful for!).

It looks like even someone like me who isn't familiar with the internals shouldn't have too much trouble adding another parameter, so if I remember (which is a big if), I'll try to formulate a patch, at which point it's more a product discussion of whether we want to keep adding arguments to the module than an issue of carving out some time to get someone to deal with it.

**Comment by eberhm** [comment]
+1

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by tima** [comment]
wontfix

Ansible errs on the side of safety and recoverability whenever possible. The synchronize module forces `–-delay-updates` to avoid leaving a destination in a broken in-between state if the underlying `rsync` process encounters an error. Those synchronizing large numbers of files that are willing to trade safety for performance should call `rsync` directly. It's nothing we want to encourage in a module like this.

**Comment by 0utsider** [comment]
What a stupid approach - rather than give a people choice - you impose upon your solution. 
The end is such that every write your module itself.

Congratulation !

**Comment by pniederlag** [comment]
+1 @0utsider @tima it's ridiculous to claim to be smarter than the users that ask for this option.

rsync has this option for some reason and it would be a more generous decision to give users of ansible the same freedom as rsync does.

If you don't have the time to support this change request in a sane manner: your point
If you claim declining this change-request is helping users of ansible in terms of security or whatsoever, I think you are plain wrong.
